### ISSUE TITLE
Flatten circuits containing CircuitOperation

### ISSUE THREAD

**Issue by mpharrigan** [quantumlib/Cirq#3783]
I want to use `CircuitOperation` to carefully construct a circuit while respecting moment structure. At some point I need to flatten/unnest it to a normal two-dimensional circuit.

@Strilanc suggests

```python
def flatten(obj: Union[cirq.Moment, cirq.Circuit, cirq.FrozenCircuit]) -> cirq.Circuit:
    if isinstance(obj, cirq.Moment):
        flattenable = [op for op in obj if isinstance(op, cirq.CircuitOperation)]
        if not flattenable:
            return cirq.Circuit(obj)
        return cirq.Circuit.zip(
            cirq.Circuit(obj - flattenable),
            *[flatten(cast(cirq.CircuitOperation, e).circuit) for e in flattenable]
        )
    return cirq.Circuit(
        [moment for e in obj for moment in flatten(e)]
    )
```

but I'm running into problems with `circuit_operation.circuit` respecting a qubit mapping.

**Comment by maffoo** [comment]
On the original `CircuitOperation` PR I [suggested](https://github.com/quantumlib/Cirq/pull/3580#discussion_r544578017) that we should add a method to "apply" the various maps (qubit, measurement, parameter) to produce a new circuit operation with maps applied. That would then work with the code from @Strilanc. Currently, this map application only happens implicitly during decomposition. I'd be happy to take a crack at this.

**Comment by mpharrigan** [comment]
Closed by #4188 if you want a PR instead of a commit hash :)
